crates/runtime/src/mmap_vec.rs

```diff
@@ -164,51 +156,23 @@ impl DerefMut for MmapVec {
     }
 }
 
-/// Helper struct to implement the `WritableBuffer` trait from the `object`
-/// crate.
-///
-/// This enables writing an object directly into an mmap'd memory so it's
-/// immediately usable for execution after compilation. This implementation
-/// relies on a call to `reserve` happening once up front with all the needed
-/// data, and the mmap internally does not attempt to grow afterwards.
-#[derive(Default)]
-struct ObjectMmap {
-    mmap: Option<MmapVec>,
-    len: usize,
-    err: Option<Error>,
-}
-
-impl WritableBuffer for ObjectMmap {
-    fn len(&self) -> usize {
-        self.len
-    }
-
-    fn reserve(&mut self, additional: usize) -> Result<(), ()> {
-        assert!(self.mmap.is_none(), "cannot reserve twice");
-        self.mmap = match MmapVec::with_capacity(additional) {
-            Ok(mmap) => Some(mmap),
-            Err(e) => {
-                self.err = Some(e);
-                return Err(());
-            }
-        };
-        Ok(())
-    }
-
-    fn resize(&mut self, new_len: usize) {
-        // Resizing always appends 0 bytes and since new mmaps start out as 0
-        // bytes we don't actually need to do anything as part of this other
-        // than update our own length.
-        if new_len <= self.len {
-            return;
-        }
-        self.len = new_len;
-    }
-
-    fn write_bytes(&mut self, val: &[u8]) {
-        let mmap = self.mmap.as_mut().expect("write before reserve");
-        mmap[self.len..][..val.len()].copy_from_slice(val);
-        self.len += val.len();
+/// Owned representation of the backing file for a `MmapVec`
+//
+// As a safety implementation-note this type cannot give access to the
+// underlying `Mmap` itself since that would violate the `DerefMut for MmapVec`
+// implementation. Instead all this gives access to is the file itself which
+// should be safe for working with in syscalls and such.
```

Comment:
Could we refactor this to hold an `Arc<File>` instead, to avoid the slightly subtle manual-mutual-exclusion note here?